controllers/actions.github.com/autoscalingrunnerset_controller.go

```diff
@@ -260,6 +247,44 @@ func (r *AutoscalingRunnerSetReconciler) Reconcile(ctx context.Context, req ctrl
 		return ctrl.Result{}, nil
 	}
 
+	if desiredSpecHash != latestRunnerSet.Labels[labelKeyRunnerSpecHash] {
+		if r.drainingJobs(&latestRunnerSet.Status) {
```

Comment:
Hey @Link-! I guess this might result in an upgrade never being completed when the old listener continuously received new workflow jobs... right?

If that's the case, can we instead do the following:

- Create a new listener immediately
- Delete the old listener (Can this result in the old ephemeralrunnerset to never drain? If that's the case this won't work...
- Wait for the latest ephemeralrunnerset to drain
- Delete the old ephemeralrunnerset

The idea is to prepare the new ephemeralrunnerset and the listener first so that the new workflow jobs immediately go to new runners, where workflow jobs that are already running on old runners are unaffected.

Reply:
> Thinking about your approach, I just realized another way we can implement it. We could extend the EphemeralRunnerSet CRD to contain the number of old ephemeral runners running in a different ephemeral runner set. That one could be controlled by the autoscaling runner set. On status update of the old ephemeral runner set, autoscaling runner set is going to be notified, and it can issue the patch to the newest ephemeral runner set and the replacement in that case would be faster.

@nikola-jokic - I'm afraid we'll be opening pandora's box here for so much unexpected behaviour. Let's keep it simple and stupid for now, see how it performs in the field and re-iterate based on that.